syntax = "proto3";

package fourx.proto;

message PlayerJoin {
  string name = 1;
  // Do not provide id on request. Use it from the SimulationSync.
  int32 id = 2;
}

message SimulationStepRequest {
  // Leave blank if unknown.
  int32 player_id = 1;
  oneof step {
    PlayerJoin player_join = 2;
  }
}

message PlayerJoinResponse {
  int32 player_id = 1;
}

// Some steps may have response data.
message SimulationStepResponse {
  oneof response {
    PlayerJoinResponse player_join_response = 1;
  }
}

message SimulationSyncRequest {
  int32 player_id = 1; 
}

message SimulationSyncResponse {
  repeated SimulationStepRequest steps = 1;
}

service SimulationService {
  rpc Step(SimulationStepRequest) returns (SimulationStepResponse) {}
  rpc Sync(SimulationSyncRequest) returns (SimulationSyncResponse) {}
}
